File: php/libs/RedisConsumer.php
<?php

namespace libs;

use libs\RedisConnect;
use DateTime;

class RedisConsumer 
{
    private RedisConnect $redis;
    private $_onMsg;
    private string $_topic;
    private string $_name;
    private string $_topicContainer;
    private string $_queueData;
    private bool $_isStart = false;

    private int $_numOfItemDequeue=2;

    private RedisConnect $redisForDequeue;

    function __construct(RedisConnect $redis, string $consumerName, string $topic, callable $onMsg)
    {
        $this->_topic = $topic;
        $this->_name = $consumerName;
        $this->_topicContainer = $topic . ":consumers";
        $this->_queueData = $topic . ":data:" . $consumerName;
        $this->redis = $redis;

        $this->redisForDequeue = new RedisConnect($redis->host, $redis->port, $redis->password, $redis->defaultDb);

        $this->_onMsg = $onMsg;
    }

    private function Do()
    {
        $counterNoMsg = 0;
        while (true) {
            try {
                $temp = [];
                for ($i = 0; $i < $this->_numOfItemDequeue; $i++) {
                    $msg = $this->redisForDequeue->Dequeue($this->_queueData);
                    if (!empty($msg)) {
                        $temp[] = $msg;
                    }
                }

                if (count($temp) > 0) {
                    foreach ($temp as $msg) {
                        call_user_func_array($this->_onMsg, [$msg]);
                    }
                } else {
                    $len = $this->redisForDequeue->ListLength($this->_queueData);

                    if ($len <= 0) {
                        $counterNoMsg = $counterNoMsg + 1;
                        usleep(10000);
                        if ($counterNoMsg > 5) {
                            $counterNoMsg = 0;
                            break;
                        }
                    } else {
                        $counterNoMsg = 0;
                    }
                }

                usleep(1000);
            } catch (\Throwable $th) {
                //
                echo "Error Do: " . $th;
                break;
            }
        }
    }

    public function Publish(string $msg)
    {
        $listConsumer =  $this->redisForDequeue->HashGetAll($this->_topic . ":consumers");

        if (count($listConsumer) == 0) return 0;

        foreach ($listConsumer as $k=>$v) {
            $queueName = $this->_topic . ":data:" . $k;
            $this->redisForDequeue->Enqueue($queueName, $msg);
        }

        return $this->redisForDequeue->Publish($this->_topic, $msg);
    }

    public function Start()
    {
        if ($this->_isStart) return;

        $this->_isStart = true;
        $date = new DateTime();
        $this->redisForDequeue->HashSet($this->_topicContainer, $this->_name, $date->format('Y-m-d H:i:s'));

        $this->Do();
        $this->redis->Subscribe($this->_topic, function ($msg) {
            //pub sub trigger dequeue
            $this->Do();
        });
    }

    public function Stop()
    {
        $this->redisForDequeue->HashDelete($this->_topicContainer, $this->_name);
        $this->redis->Unsubscribe($this->_topic);        
        $this->Do();
        $this->_isStart=false;
    }

    function __destruct()
    {
        $this->Stop();
    }
}

File: RedisAsMessageBroker.Core/RedisConsumer.cs
using StackExchange.Redis;

namespace RedisAsMessageBroker.Core
{
    public class RedisConsumer
    {
        readonly RedisConnect _redis;

        readonly string _topic;

        readonly string _name;

        readonly string _queueData;

        readonly Func<string, Task> _onMsg;

        readonly string _topicContainer;

        ISubscriber _subscriber;
        ChannelMessageQueue _channelMsgQueue;

        bool _started = false;
        readonly int _numOfItemDequeue = (Environment.ProcessorCount / 4)+1;
        public RedisConsumer(RedisConnect redis, string consumerName, string topic, Func<string, Task> onMsg)
        {
            _redis = redis;
            _topic = topic;
            _name = consumerName;

            _queueData = $"{_topic}:data:{_name}";

            _topicContainer = $"{topic}:consumers";

            _onMsg = onMsg;

        }

        async Task Do()
        {
            int counterNoMsg = 0;
            while (true)
            {
                try
                {
                    List<string> temp = new List<string>();
                    for (var i = 0; i < _numOfItemDequeue; i++)
                    {
                        var msg = await _redis.Dequeue(_queueData);

                        if (!string.IsNullOrEmpty(msg))
                            temp.Add(msg);
                    }
                    if (temp.Count > 0)
                    {
                        List<Task> tasks = new List<Task>();
                        foreach (var t in temp)
                        {
                            tasks.Add(_onMsg(t));
                        }
                        await Task.WhenAll(tasks);
                    }
                    else
                    {
                        var lng = await _redis.ListLength(_queueData);
                        if (lng <= 0)
                        {
                            counterNoMsg++;
                            if (counterNoMsg > 5)
                            {
                                counterNoMsg = 0;
                                break;
                            }
                            await Task.Delay(100);
                        }
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine(ex);
                    break;
                }

                await Task.Delay(1);
            }

        }

        async Task<ISubscriber> GetSubscriber()
        {
            _subscriber = _subscriber ?? await _redis.GetSubscriber();

            return _subscriber;
        }
        public async Task<long> Publish(string msg)
        {
            var listConsumer = await _redis.HashGetAll(_topicContainer);

            if (listConsumer.Count <= 0) return 0;

            List<Task> tasks = new List<Task>();
            foreach (var c in listConsumer)
            {
                var queueName = $"{_topic}:data:{c.Key}";
                tasks.Add(_redis.Enqueue(queueName, msg));
            }

            await Task.WhenAll(tasks);

            return await (await GetSubscriber()).PublishAsync(_topic, msg);
        }
        public async Task Start()
        {
            await Task.Yield();

            if (_started) return;

            _started = true;

            await _redis.HashSet(_topicContainer, _name, DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss"));

            _ = Do();

            _channelMsgQueue = await (await GetSubscriber()).SubscribeAsync(_topic);

            _channelMsgQueue.OnMessage((msg) =>
            {
                //pub sub trigger dequeue
                _ = Do();
            });

          
        }

        public async Task Stop()
        {
            await _redis.HashDelete(_topicContainer, _name);
            await (await GetSubscriber()).UnsubscribeAsync(_topic);
            _ = Do();

            _started = false;
        }
    }
}